Lazuri - first submission

makerling opened the pull request:
Hi, this is my first try at uploading a keyboard to the Keyman repository.  Let me know if there are any issues that need addressing. Thanks!

DavidLRowe commented:
Keyboard looks good, producing desired keystrokes on desktop (and to a limited degree on touch -- see comment below).

Things to address before release:
- Copyright statment in README.md and .kmn file should include 2017 (as it does in LICENSE.md) or date range if keyboard existed before 2017.
- README.md and HISTORY.md have version as 1.0, .kmn file has version 2.0. They all need to agree.
- project file (.kpj file) should set build path to $PROJECTPATH\build
- project file (.kpj file) should set CompilerWarningsAsErrors to True

Other things to consider:
- Since there is no lazuri-layout.js file (to define a touch layout), any touch layout will be based on the desktop layout, which means Alt and Ctrl+Alt combinations from the .kmn file won't be accessible. You may want to consider creating a touch layout for use on phone/tablet that will provide access to those keys.
- very minor: welcome.htm doesn't have utf-8 spec. This isn't neeeded at the moment, but helpful to include.
- There are duplicate copies of Lazuri_Lazca_Q_Klavye.pdf (one in source/welcome, one in docs/)
- herdilder.bmp doesn't appear to be used anywhere; is it intended for inclusion in welcome.htm?

DavidLRowe commented:
I think the failure to build on Team City is due to the build path issue noted above.

DavidLRowe commented:
BTW, the original .kpj (project file) didn't actually seem to reference the .kmn file. That also might cause a build failure!

makerling commented:
Thanks, I'll work on fixing the issues.

mcdurdin commented on experimental/l/lazuri/lazuri.keyboard_info:
```diff
@@ -0,0 +1,28 @@
+        "windows": "full",
+        "ios": "basic",
+        "macos": "full"
+    },
```
This trailing comma should be removed to make the JSON file valid.

makerling commented:
OK, I think I've fixed the issues addressed above. I've removed the touch layouts, as we haven't finished them. There are duplicate copies, I wanted one to be referenced with a relative link from the welcome.htm file. I wasn't sure how to do that without adding it to the docs folder. I've associated the herdilder.bmp in the .kps file. I didn't know I needed to add the .kmn file to the .kps file, the documentation says to add the .kmx. I've added it, hopefully that helps. I've remove the stray comma in the lazuri.keyboard_info file. Hopefully I didn't create new problems with my changes : ). Can you let me know if the build succeeds? 

Thanks!

DavidLRowe commented:
Under "All checks have failed", you should see the one failing check, namely the TeamCity continuous integration build. To check the build status, click on the "Details" link. This should invite you to "Log in to TeamCity". Click the "Log in as guest" (unless, of course, you have a TeamCity login!) and it should show you the build problems.

```
Process exited with code 1
[14:08:28]cp: cannot stat 'source/lazuri.kmp': No such file or directory
[14:08:28]
[14:08:28]Aborting with error 1
```

This indicates it is looking for the lazuri.kmp file in the `source` folder. (I'm not sure why that is.) It fails because that file isn't in the `source` folder (but it shouldn't be there; it should be in `build`).

In the welcome.htm file, the links to the pdf need to change to match the changed pdf filename.

The .kps file should reference:
```
..\build\lazuri.kmx
..\build\lazuri.kvk
welcome\welcome.htm
welcome\Lazca_Lazuri_Q_Klavye.pdf
locale-tur.xml
```
Note that the .kvk file to reference is the one in the **build** folder (not the `source` folder), and also that paths are relative to the `source` folder, since that is where the .kps file is located.

The .kpj file should reference the files referenced in the .kps file (but paths are relative to the project folder, since that is where the .kpj file is located), plus additional files (the .kmn file, the .ico file referenced by the .kmn, and the .kps file):
```
source\..\build\lazuri.kmx
source\..\build\lazuri.kvk
source\welcome\welcome.htm
source\welcome\Lazca_Lazuri_Q_Klavye.pdf
source\locale-tur.xml
source\lazuri.ico
source\lazuri.kmn
source\lazuri.kps
```

The .kpj doesn't need a reference to README.md. (Note that this is not the same file as the optional readme.htm file, which can be included for display when the keyboard in installed.)

If the herdilder.bmp is referenced by welcome.htm, then it would need to be included in both the .kps and .kpj files.

mcdurdin commented:
Note: currently build status for keyboard builds is not public; we need to complete an audit on it before we can make it public.

MakaraSok commented:
@makerling @DavidLRowe It's not convenient to work off of others' master branch. One can only pull from the branch, but no permission to push any update to it. Is there anyway to push updates to a master branch?

makerling commented:
Thanks. I made the changes above. I checked the build status and noticed it failed with the same error as last time "cp: cannot stat 'source/lazuri.kmp': No such file or directory". Any ideas what might be causing this?

MakaraSok commented:
@makerling you may not have included the .kmn in your .kpj.

makerling commented:
The kmn is included/referenced in my .kpj file:

[https://github.com/makerling/keyboards/blob/master/experimental/l/lazuri/lazuri.kpj](url)

`    <File>
      <ID>id_66703b22c8e0d5f23e1eeac196cfd9d8</ID>
      <Filename>lazuri.kmn</Filename>
      <Filepath>source\lazuri.kmn</Filepath>
      <FileVersion>1.0</FileVersion>
      <FileType>.kmn</FileType>
      <ParentFileID>id_fd7892f21900684c8589f87337c4f2f6</ParentFileID>
      <Details>
        <Name>Lazca (Lazuri) Klavye</Name>
        <Copyright>© 2017 HERdilDER</Copyright>
        <Message>Lazca (Lazuri) Klavye</Message>
      </Details>
    </File>
    <File>`

MakaraSok commented on experimental/l/lazuri/source/lazuri.kps:
```diff
@@ -0,0 +1,106 @@
+    </File>
+    <File>
+      <Name>lazuri.kmn</Name>
+      <Description>File lazuri.kmn</Description>
```
@makerling .kmn should not be in the .kps. Let's remove it.

MakaraSok commented on experimental/l/lazuri/source/lazuri.kps:
```diff
@@ -0,0 +1,106 @@
+      <FileType>.kvk</FileType>
+    </File>
+    <File>
+      <Name>..\README.md</Name>
```
this file should also be removed. If you have README.htm, then include it; but let's not include README.md.

MakaraSok replied:
@makerling

makerling commented:
Thanks @MakaraSok. I updated the .kps file with your recommendations. However, the build failed again with the same error : (.

MakaraSok commented on experimental/l/lazuri/source/lazuri.kps:
```diff
@@ -0,0 +1,94 @@
+      <FileType>.bmp</FileType>
+    </File>
+    <File>
+      <Name>locale-tur.xml</Name>
```
@makerling it looks like there are two copies of this file in here.

MakaraSok commented on experimental/l/lazuri/source/lazuri.kmn:
```diff
@@ -0,0 +1,134 @@
+store(&VERSION) '9.0'
+store(&NAME) 'Lazca (Lazuri) Klavye'
+store(&TARGETS) 'windows macosx linux'
+store(&ETHNOLOGUECODE) 'Lzz'
```
'Lzz' should be all small caps.

MakaraSok replied:
@makerling I think the language code you have in .keyboard_info is all lower case. Let's make sure these two are consistent. Not sure if it fixes the build error though.

MakaraSok commented:
@DavidLRowe do you know why this happens? the .keyboard_info file is valid according to the JSONLint, but the build fails anyway.

@makerling could you please update your Keyman Developer? Please find the latest version [here](https://keyman.com/downloads/pre-release/). At the time of my visit, the latest KD alpha is `keymandeveloper-10.0.977.0.exe alpha (released 2017-12-13, 25.14 MB)`. 

Then please re-compile the .kmn and .kps, push any changes of the keyboard up, and let's see if the build still fails.

Thanks!

mcdurdin commented:
I can take a look at this in the New Year (our Keyman team office is closed between now and 1 Jan)

mcdurdin commented:
Okay, the issue here, which I should have been able to spot earlier, is that keyboards in the `experimental` folder are not currently built in the same way as keyboards in the `release` folder. Is there any reason we need the keyboard to stay in `experimental`?

Note that the experimental category is not yet supported in Keyman products -- this needs to happen prior to the release of Keyman 10 on all platforms -- and so there is no clear immediate benefit to publishing under that category instead of under `release`.

Recommendation: move the keyboard folder to `release`. I have tested this and it builds successfully under `release`.

MakaraSok commented:
@makerling If you expect to release this keyboard, let's move it to `release` under `l` directory (if `l` directory doesn't exist, feel free to create one). After it's moved, this PR should be closed.

mcdurdin commented:
No need to close the PR. Move to a new folder and push.

On Thu, 11 Jan 2018 at 2:58 pm, Makara <notifications@github.com> wrote:

> @makerling <https://github.com/makerling> If you expect to release this
> keyboard, let's move it to release under l directory (if l directory
> doesn't exist, feel free to create one). After it's moved, this PR should
> be closed.
>
> —
> You are receiving this because you commented.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/keymanapp/keyboards/pull/130#issuecomment-356817853>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AESjvUuPpb_nc4bFL1I1M5o9TxkL-ZJVks5tJYb1gaJpZM4Q0naa>
> .
>

makerling commented:
OK, thanks for the advice. I've moved it from experimental to release and the build passed. It would be helpful if the documentation ([https://help.keyman.com/developer/keyboards/](url)) could be updated if the experimental folder shouldn't be used, so the next person who tries it can succeed. Thanks.

mcdurdin commented:
I've added a todo item to update the web page or the build script to resolve the inconsistency.

makerling commented:
Thanks. I have a question about the merge date for the ready-to-merge tagged keyboards. Is there an expected date when this and the 11 other ready-to-merge pull requests will be merged? Also, I have a few edits I'd like to make on this keyboard, should I wait until the merge happens and then submit another pull request, or can I do the edits to this existing pull request?j Thanks!

DavidLRowe commented:
At the moment the merging process (after the github merge is done) requires some manual intervention to get the merged keyboards in the right place. Work is being done to automate this process, but sadly this means there's been a delay in actually merging keyboards. 

If you'd like to make additional changes, we can remove the "ready-to-merge" label and cancel the review. If the changes are small, it shouldn't take much to update the review.

mcdurdin commented:
Published keyboard:
 - Home: https://keyman.com/keyboard/lazuri
 - Legacy: http://keymankeyboards.com/?id=725
 - Documentation: https://help.keyman.com/keyboard/lazuri/

makerling commented:
Many thanks to all your help in getting this published!
